File: notebooksjava/_01IsUnique.java
import java.util.HashSet;
import java.util.Set;

public class _01IsUnique {

    // 解法1：HashSet — O(n) time, O(n) space
    public boolean isUnique(String astr) {
        Set<Character> seen = new HashSet<>();
        for (char c : astr.toCharArray()) {
            if (!seen.add(c)) return false;
        }
        return true;
    }

    // 解法2：ビット演算 — O(n) time, O(1) space（a-z のみ対応）
    public boolean isUniqueBit(String astr) {
        int mark = 0;
        for (char c : astr.toCharArray()) {
            int bit = c - 'a';
            if ((mark & (1 << bit)) != 0) return false;
            mark |= (1 << bit);
        }
        return true;
    }

    public static void main(String[] args) {
        _01IsUnique sol = new _01IsUnique();

        // isUnique
        assert sol.isUnique("abc")      == true  : "abc should be unique";
        assert sol.isUnique("apple")    == false : "apple has duplicate 'p'";
        assert sol.isUnique("")         == true  : "empty string is unique";
        assert sol.isUnique("a")        == true  : "single char is unique";
        assert sol.isUnique("leetcode") == false : "leetcode has duplicates";

        // isUniqueBit
        assert sol.isUniqueBit("abc")      == true  : "abc should be unique (bit)";
        assert sol.isUniqueBit("apple")    == false : "apple has duplicate 'p' (bit)";
        assert sol.isUniqueBit("leetcode") == false : "leetcode has duplicates (bit)";

        System.out.println("All test cases passed ✅");
    }
}

File: notebookspython/01-is-unique.py
class SolutionSet:
    def isUnique(self, astr: str) -> bool:
        return len(astr) == len(set(astr))


class Solution:
    def isUnique(self, astr: str) -> bool:
        """Check uniqueness for lowercase a-z characters using a bit mask."""
        mark = 0

        for char in astr:
            move_bit = ord(char) - ord("a")
            if (mark & (1 << move_bit)) != 0:
                return False
            mark |= 1 << move_bit

        return True


if __name__ == "__main__":
    sol = Solution()

    assert sol.isUnique("abc") is True
    assert sol.isUnique("leetcode") is False
    assert sol.isUnique("algorithm") is True
    assert sol.isUnique("hello") is False
    assert sol.isUnique("") is True

    print("All test cases passed ✅")
